src/lambda/rest-api/PersonaAuthorizer.ts

```diff
@@ -0,0 +1,310 @@
+    stage: string;
+  };
+  awsAccountId: string;
+};
+
+// Constants used by parseMethodArn:
+//
+// Example MethodARN:
+//   "arn:aws:execute-api:<Region id>:<Account id>:<API id>/<Stage>/<Method>/<Resource path>"
+// Method ARN Index:  0   1   2           3           4            5
+// API Gateway ARN Index:                                          0        1       2        3
+//
+//
+const ARN_INDEX = 0;
+const AWS_INDEX = 1;
+const EXECUTE_INDEX = 2;
+const REGION_INDEX = 3;
+const ACCOUNT_ID_INDEX = 4;
+const API_GATEWAY_ARN_INDEX = 5;
+
+const METHOD_ARN_INDEXES = [
+  ARN_INDEX,
+  AWS_INDEX,
+  EXECUTE_INDEX,
+  REGION_INDEX,
+  ACCOUNT_ID_INDEX,
+  API_GATEWAY_ARN_INDEX,
+];
+
+const API_ID_INDEX = 0;
+const STAGE_INDEX = 1;
+const METHOD_INDEX = 2;
+const RESOURCE_PATH_INDEX = 3;
+
+const API_GATEWAY_ARN_INDEXES = [
+  API_ID_INDEX,
+  STAGE_INDEX,
+  METHOD_INDEX,
+  RESOURCE_PATH_INDEX,
+];
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+export class PersonaAuthorizer {
+  event: any;
+  context: any;
+  personaClient: PersonaClient | undefined;
+
+  constructor(event: any, context: any) {
+    this.event = event;
+    this.context = context;
+
+    this.personaClient = undefined;
+  }
+
+  async handle() {
+    console.log("Received event", this.event);
+
+    if (
+      !this.event?.authorizationToken ||
+      this.event.authorizationToken == null
+    ) {
+      console.log("Missing auth token");
+      return this.context.fail("Unauthorized");
+    }
+
+    const parsedMethodArn = this.parseMethodArn(this.event.methodArn);
+    console.log(`Parsed Method Arn: ${JSON.stringify(parsedMethodArn)}`);
+
+    const scope = this.getScope(parsedMethodArn);
+    console.log(`Method has scope: ${scope}`);
+
+    let validationOpts = {
+      token: _.replace(this.event.authorizationToken, "Bearer", "").trim(),
+    };
+    if (scope != null) {
+      validationOpts = _.merge(validationOpts, { scope });
+    }
+    console.log(`Validation ops: ${JSON.stringify(validationOpts)}`);
+
+    console.log(
+      "validating token against request",
+      `${parsedMethodArn.resourcePath}`,
+    );
+
+    if (!validationOpts.token || validationOpts.token.length === 0) {
+      console.log("token missing");
+      return this.context.fail("Unauthorized");
+    }
+
+    try {
+      const decodedToken = await this.validateToken(validationOpts);
+      const principalId = _.get(decodedToken, "aud", "invalid_oauth");
+      const authPolicy = this.buildAuthPolicy(
+        principalId,
+        parsedMethodArn,
+        true,
+      );
+      return this.context.succeed(authPolicy);
+    } catch (err) {
+      console.log("token validation failed", err);
+
+      // In the case of suceess - the principal id is coming from the
+      // decoded token. We don't have it here for the case of an invalid token.
+      // Leaving the proncipal id blank in the auth policy for deny for now.
+      // But could this be found?
```

Comment:
What does this mean?

Reply:
Great question. Probably one for CC - I guess this has always been here. Happy to delete the comment.